app/uk/gov/hmrc/agentsubscriptionfrontend/controllers/SubscriptionController.scala

```diff
@@ -130,56 +129,67 @@ class SubscriptionController @Inject()
+    }
+  }
 
-        def redirectSubscriptionResponse(either: Either[SubscriptionReturnedHttpError, (Arn, String)]): Result = {
-          either match {
-            case Right((arn, agencyName))                      => Redirect(routes.SubscriptionController.showSubscriptionComplete())
-              .flashing("arn" -> arn.arn, "agencyName" -> agencyName)
-            case Left(SubscriptionReturnedHttpError(CONFLICT)) => Redirect(routes.CheckAgencyController.showAlreadySubscribed())
-            case Left(SubscriptionReturnedHttpError(_))        => Redirect(routes.SubscriptionController.showSubscriptionFailed())
-          }
-        }
+  def returnFromAddressLookup(id: String): Action[AnyContent] = AuthorisedWithSubscribingAgentAsync() {
+    implicit authContext =>
+      implicit request =>
 
         sessionStoreService.fetchInitialDetails.flatMap { maybeDetails =>
           maybeDetails.map { details =>
             addressLookUpConnector.getAddressDetails(id).flatMap { address =>
-              addressLookUpValidator.validateAddress(details.utr, address, blacklistedPostCodes) match {
-                case Invalid(errors)   =>
-                  Future.successful(
-                    Ok(des_will_not_accept_address(id, SubscriptionController.renderErrors(errors)))
-                  )
-                case Valid(desAddress) =>
-                  val subscriptResponse = for {
-                    res ← subscribe(details, desAddress, address)
-                  } yield res
-                  subscriptResponse.map(redirectSubscriptionResponse)
+              val addressForm = desAddressForm.bindAddressLookupFrontendAddress(details.utr, address)
+              if (addressForm.hasErrors) {
+                Future successful
+                  Ok(html.address_form_with_errors(addressForm))
+              } else {
+                subscribe(details, addressForm.get, Some(address)).map(redirectSubscriptionResponse)
               }
             }
           }.getOrElse(Future.successful(sessionMissingRedirect()))
         }
   }
 
-  def restartAddressLookup(id: String): Action[AnyContent] = AuthorisedWithSubscribingAgentAsync() {
+  def submitModifiedAddress: Action[AnyContent] = AuthorisedWithSubscribingAgentAsync() {
     implicit authContext =>
       implicit request =>
-
-        addressLookUpConnector.initJourney(routes.SubscriptionController.returnFromAddressLookup(), JourneyName).map { x => Redirect(x) }
+        desAddressForm.form.bindFromRequest().fold(
+          formWithErrors => {
+            Future successful Ok(html.address_form_with_errors(formWithErrors))
+          },
+          validDesAddress =>
+            sessionStoreService.fetchInitialDetails.flatMap { maybeInitialDetails =>
+              maybeInitialDetails.map { initialDetails =>
+                val subscriptResponse = for {
+                  res ← subscribe(initialDetails, validDesAddress, None)
+                  _ ← sessionStoreService.remove()
```

Comment:
Looks like this call to `sessionStoreService.remove()` might break the continueUrl functionality - it relies on the continue URL being available in `showSubscriptionComplete` which happens after this.